# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
import pandas as pd
import cleaning_functions as cf

# Cargar los datasets
file1 = pd.read_csv('data/raw/file1.csv')
file2 = pd.read_csv('data/raw/file2.csv')
file3 = pd.read_csv('data/raw/file3.csv')

# Renombrar columnas para que coincidan
file1.rename(columns={'ST': 'State', 'GENDER': 'Gender'}, inplace=True)
file2.rename(columns={'ST': 'State', 'GENDER': 'Gender'}, inplace=True)

# Reordenar columnas de file3 para que coincidan con file1 y file2
file3 = file3[['Customer', 'State', 'Gender', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']]

# Reordenar columnas de file1 y file2 para que coincidan con file3
file1 = file1[['Customer', 'State', 'Gender', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']]
file2 = file2[['Customer', 'State', 'Gender', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']]

# Combinar los datasets
combined_data = pd.concat([file1, file2, file3], ignore_index=True)

# Función para limpiar y estandarizar la columna 'Gender'
def clean_gender(gender):
    if pd.isna(gender):
        return gender
    gender = gender.strip().lower()
    if gender in ['m', 'male']:
        return 'M'
    elif gender in ['f', 'female', 'femal']:
        return 'F'
    else:
        return gender

# Aplicar la función a la columna 'Gender'
combined_data['Gender'] = combined_data['Gender'].apply(clean_gender)


# Función para limpiar y estandarizar la columna 'State'
def clean_state(state):
    if pd.isna(state):
        return state
    state = state.strip().lower()
    state_mapping = {
        'cali': 'California',
        'california': 'California',
        'az': 'Arizona',
        'arizona': 'Arizona',
        'nv': 'Nevada',
        'nevada': 'Nevada',
        'or': 'Oregon',
        'oregon': 'Oregon',
        'wa': 'Washington',
        'washington': 'Washington'
    }
    return state_mapping.get(state, state.capitalize())

# Aplicar la función a la columna 'State'
combined_data['State'] = combined_data['State'].apply(clean_state)

# Calcula los null.sum de cada columna
print(f"Número de nulos a tratar: {combined_data.isnull().sum()}")

# Calcula los duplicated.sum de cada columna
print(f"Número de duplicados a tratar: {combined_data.duplicated().sum()}")

# Eliminar duplicados
combined_data.drop_duplicates(inplace=True)

# Limpiar y estandarizar el dataset
combined_data = cf.clean_format_main(combined_data)

# Guardar el dataset combinado
combined_data.to_csv('data/processed/combined_data.csv', index=False)

Número de nulos a tratar: Customer                     2937
State                        2937
Gender                       3059
Education                    2937
Customer Lifetime Value      2944
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64
Número de duplicados a tratar: 2939
gender: [nan 'F' 'M']
education: ['Master' 'Bachelor' 'High School or Below' 'College' 'Doctor' nan]
state: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' nan]
Columns with null or missing values:
Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')
All null values have been successfully handled
0.0


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [ ]:
import pandas as pd
import cleaning_functions as cf

# Cargar el dataset
marketing_data = pd.read_csv('data/raw/marketing_customer_analysis_clean.csv')

# Mostrar las primeras filas del dataset para inspección
print(marketing_data.head())

# Limpiar y formatear los datos según sea necesario
# Por ejemplo, eliminar columnas innecesarias, manejar valores nulos, etc.

# Ejemplo: Eliminar columnas innecesarias
columns_to_drop = ['unnamed:_0']
marketing_data.drop(columns=columns_to_drop, inplace=True)

# Ejemplo: Manejar valores nulos
marketing_data.ffill(inplace=True)

# Identificar las columnas categóricas
categorical_columns = marketing_data.select_dtypes(include=['object']).columns
# print("Columnas categóricas:", categorical_columns)

# Revisar los valores únicos en las columnas categóricas
# for column in categorical_columns:
    # print(f"Valores únicos en '{column}':")
    # print(marketing_data[column].unique())
    # print()

# Crear una función para homogeneizar la columna 'education'
def clean_education(education):
    if pd.isna(education):
        return education
    education = education.strip().lower()
    education_mapping = {
        'high school or below': 'High School or Below',
        'bachelor': 'Bachelor',
        'college': 'College',
        'master': 'Master',
        'doctor': 'Doctor'
    }
    return education_mapping.get(education, education.capitalize())

# Aplicar la función a la columna 'education'
marketing_data['education'] = marketing_data['education'].apply(clean_education)

# Elimina duplicados
marketing_data.drop_duplicates(inplace=True)

marketing_data = cf.clean_format_main(marketing_data)

# Guardar el dataset limpio
marketing_data.to_csv('data/processed/marketing_customer_analysis_cleaned.csv', index=False)

marketing_data['education'].unique()



   unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective_to_date employmentstatus gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3          

array(['College', 'Bachelor', 'High School or Below', 'Doctor', 'Master'],
      dtype=object)

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [5]:
import pandas as pd

# Cargar el dataset limpio
marketing_data = pd.read_csv('data/processed/marketing_customer_analysis_cleaned.csv')

# Crear una tabla pivote para resumir los ingresos totales por canal de ventas
summary_table = marketing_data.pivot_table(
    values='total_claim_amount', 
    index='sales_channel', 
    aggfunc='sum'
).reset_index()

# Redondear los ingresos totales a 2 decimales
summary_table['total_claim_amount'] = summary_table['total_claim_amount'].round(2)

# Mostrar la tabla resumen
print(summary_table)

  sales_channel  total_claim_amount
0         Agent             1742958
1        Branch             1255658
2   Call Center              886959
3           Web              675009


### Insights

- Agent Channel: The "Agent" sales channel has the highest total revenue, amounting to 1,742,958. This indicates that customers who interact with agents tend to generate the most revenue.
- Branch Channel: The "Branch" sales channel is the second highest in terms of total revenue, with 1,255,658. This suggests that physical branch interactions are also significant contributors to revenue.
- Call Center Channel: The "Call Center" sales channel has a total revenue of 886,959, making it the third highest. This shows that call centers are an important but less dominant sales channel compared to agents and branches.
- Web Channel: The "Web" sales channel has the lowest total revenue, with 675,009. This indicates that online interactions generate the least revenue among the four channels.

### Recommendations

- Focus on Agent and Branch Channels: Since these channels generate the most revenue, consider investing more in training and resources for agents and branches to further enhance their performance.
- Improve Call Center Efficiency: Explore ways to improve the efficiency and effectiveness of the call center to potentially increase its revenue contribution.
- Enhance Online Presence: Investigate why the web channel generates the least revenue and consider strategies to improve the online customer experience and conversion rates.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [6]:
import pandas as pd

# Cargar el dataset limpio
marketing_data = pd.read_csv('data/processed/marketing_customer_analysis_cleaned.csv')

# Crear una tabla pivote para mostrar el valor promedio de vida del cliente por género y nivel educativo
pivot_table = marketing_data.pivot_table(
    values='customer_lifetime_value', 
    index='gender', 
    columns='education', 
    aggfunc='mean'
).round(2)

# Mostrar la tabla pivote
print(pivot_table)

education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7912.81  7756.81  7369.08               8617.29  8224.11
M           7698.13  8074.74  7414.38               8158.88  8154.10


### Insights

- Gender Comparison: On average, male customers have a slightly higher customer lifetime value compared to female customers across all education levels.
- Education Level Comparison: Customers with a "Doctor" or "Premium" education level have the highest average customer lifetime value, indicating that higher education levels are associated with higher customer lifetime values.
- High School or Below: Customers with "High School or Below" education level have the lowest average customer lifetime value, suggesting that lower education levels are associated with lower customer lifetime values.

### Recommendations

- Target Higher Education Levels: Marketing strategies could focus more on customers with higher education levels, as they tend to have higher customer lifetime values.
- Gender-Specific Campaigns: Consider creating gender-specific marketing campaigns to address the slight differences in customer lifetime value between male and female customers.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [7]:
import pandas as pd

# Cargar el dataset limpio
marketing_data = pd.read_csv('data/processed/marketing_customer_analysis_cleaned.csv')

# Extraer el mes de la columna 'effective_to_date'
marketing_data['month'] = pd.to_datetime(marketing_data['effective_to_date']).dt.month

# Crear una tabla resumen para mostrar el número de quejas por tipo de póliza y mes
summary_table = marketing_data.groupby(['policy_type', 'month'])['number_of_open_complaints'].sum().reset_index()

# Mostrar la tabla resumen en formato largo
print(summary_table)

      policy_type  month  number_of_open_complaints
0  Corporate Auto      1                        392
1  Corporate Auto      2                        347
2   Personal Auto      1                       1583
3   Personal Auto      2                       1312
4    Special Auto      1                         81
5    Special Auto      2                         82


### Insights

- Corporate Auto: The number of complaints for "Corporate Auto" policy type is higher in February compared to January.
- Personal Auto: The "Personal Auto" policy type has the highest number of complaints in both January and February, indicating that this policy type may require more attention.
- Special Auto: The "Special Auto" policy type has the lowest number of complaints, suggesting that customers with this policy type are generally more satisfied.

### Recommendations

- Focus on Personal Auto: Since the "Personal Auto" policy type has the highest number of complaints, consider investigating the reasons behind the complaints and implementing measures to improve customer satisfaction.
- Monitor Trends: Keep monitoring the number of complaints by policy type and month to identify any emerging trends and address issues proactively.